clang/lib/Sema/SemaConcept.cpp

```diff
@@ -1109,12 +1109,50 @@ bool Sema::EnsureTemplateArgumentListConstraints(
   return false;
 }
 
-bool Sema::CheckInstantiatedFunctionTemplateConstraints(
+static bool CheckFunctionConstraintsWithoutInstantiation(
+    Sema &SemaRef, SourceLocation PointOfInstantiation,
+    FunctionTemplateDecl *Template, ArrayRef<TemplateArgument> TemplateArgs,
+    ConstraintSatisfaction &Satisfaction) {
+  SmallVector<const Expr *, 3> TemplateAC;
+  Template->getAssociatedConstraints(TemplateAC);
+  if (TemplateAC.empty()) {
+    Satisfaction.IsSatisfied = true;
+    return false;
+  }
+
+  LocalInstantiationScope Scope(SemaRef);
+
+  FunctionDecl *FD = Template->getTemplatedDecl();
+  // Collect the list of template arguments relative to the 'primary'
+  // template. We need the entire list, since the constraint is completely
+  // uninstantiated at this point.
+  MultiLevelTemplateArgumentList MLTAL =
+      SemaRef.getTemplateInstantiationArgs(FD, /*DC=*/nullptr,
+                                           /*Final=*/false,
+                                           /*Innermost=*/TemplateArgs,
+                                           /*RelativeToPrimary=*/true,
+                                           /*ForConstraintInstantiation=*/true);
+
+  std::optional<Sema::CXXThisScopeRAII> ThisScope;
```

Comment:
Do we need to point CurContext to the function here as well?

Reply:
Yes, I think so - but I couldn't find a case where that affects anything.

Reply:
Better safe than sorry!

Reply:
Added a ContextRAII anyway